File: 628. Maximum Product of Three Numbers/628.go
package  main

import (
	"fmt"
	"sort"
)

func main(){
	fmt.Println(maximumProduct([]int{1,2,3}))
}

func maximumProduct(nums []int) int {
	maxint := int(^uint(0) >> 1)
	minint := ^maxint
	max1, max2, max3, min1, min2 := minint, minint, minint, maxint, maxint
	for _,value := range nums{
		if value > max1{
			max1, max2, max3 = value, max1, max2
		}else if value>max2{
			max2, max3 = value, max2
		}else if value>max3{
			max3 = value
		}else{

		}

		if value < min1{
			min1, min2 = value, min1
		}else if value < min2{
			min2 = value
		}else{

		}
	}

	if max1*max2*max3 > max1*min1*min2{
		return max1*max2*max3
	}
	return max1*min1*min2
}

// go的int最大值最小值
//maxint := int(^uint(0) >> 1)
//minint := ^maxint

//这个和python中用到的方法一样，这里导入了sort
func maximumProduct2(nums []int) int {
	sort.Ints(nums)
	lennums := len(nums)

	a := nums[0] * nums[1] * nums[lennums-1]
	b := nums[lennums-3] * nums[lennums-2] * nums[lennums-1]

	if a > b {
		return a
	}
	return b
}

File: 628. Maximum Product of Three Numbers/628.py
class Solution:
    def maximumProduct(self, nums) -> int:
        nums.sort()
        if nums[-1]<=0:
            return nums[-1]*nums[-2]*nums[-3]
        elif nums[-2]<=0:
            return nums[0]*nums[1]*nums[-1]
        elif nums[-3]<=0:
            return nums[0]*nums[1]*nums[-1]
        else:
            return max(nums[0]*nums[1]*nums[-1], nums[-1]*nums[-2]*nums[-3])

######################################################
# 我的方法整理一下其实可以简化为下面一句
    def maximumProduct1(self, nums) -> int:
        nums.sort()
        return max(nums[0]*nums[1]*nums[-1], nums[-1]*nums[-2]*nums[-3])

# 这个方法可能效率并不是很高，但用堆值得学习借鉴
    def maximumProduct2(self, nums):
        import heapq
        a, b = heapq.nlargest(3, nums), heapq.nsmallest(2, nums)
        return max(a[0] * a[1] * a[2], b[0] * b[1] * a[0])

# 这个和最开始我想用的方法差不多，只是我一直没有理清真正需要的数，和比较的数
    def maximumProduct3(self, nums):
        max1,max2,max3,min1,min2 = float('-Inf'),float('-Inf'),float('-Inf'),float('Inf'),float('Inf')
        for num in nums:
            if num >= max1:
                max3,max2,max1 = max2,max1,num
            elif num >= max2:
                max3,max2 = max2,num
            elif num > max3:
                max3 = num
            if num <= min1:
                min2,min1 = min1,num
            elif num < min2:
                min2 = num
        return max(max1*max2*max3,min1*min2*max1)

a = Solution()
print(a.maximumProduct([1,2,3,4]))
